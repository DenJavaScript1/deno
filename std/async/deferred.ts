// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.

export class Deferred<T> extends Promise<T> {
  resolve!: (value: T | PromiseLike<T>) => void;
  // deno-lint-ignore no-explicit-any
  reject!: (reason?: any) => void;

  //it's necessary to pass an executor so that
  //.then() can function properly on the returned deferred object
  constructor(
    executor?: (
      resolve: (value: T | PromiseLike<T>) => void,
      // deno-lint-ignore no-explicit-any
      reject: (reason?: any) => void,
    ) => void,
  ) {
    //For .then() to function properly
    if (executor instanceof Function) {
      return new Promise<T>(executor) as Deferred<T>;
    }

    let methods: {
      resolve: (value: T | PromiseLike<T>) => void;
      // deno-lint-ignore no-explicit-any
      reject: (reason?: any) => void;
    };

    //capture references to resolve and reject functions
    super((resolve, reject) => {
      methods = { resolve, reject };
    });

    this.resolve = methods!.resolve;
    this.reject = methods!.reject;
  }
}
