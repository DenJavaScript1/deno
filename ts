#!/usr/bin/env -S deno run
// deno-lint-ignore-file prefer-const
let message = 'Hello, Deno!';
console.log(message);