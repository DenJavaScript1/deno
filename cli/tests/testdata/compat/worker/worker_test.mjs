import { deferred } from "../../../../../test_util/std/async/deferred.ts";

const promise1 = deferred();
const promise2 = deferred();
const url = new URL("./worker.mjs", import.meta.url);
const worker = new Worker(url.href, { type: "module", deno: true });

worker.onmessage = (e) => {
    if (e.data == "hello from worker") {
        promise1.resolve();
        return;
    }

    const pid = e.data.pid;
    if (typeof pid != "number") {
        throw new Error("pid is not a number");
    }
    console.log("process.pid from worker:", pid);
    promise2.resolve();
};

await promise1;
worker.postMessage("hello");
await promise2;
worker.terminate();
