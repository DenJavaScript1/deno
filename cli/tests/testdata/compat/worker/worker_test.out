process.pid from worker: [WILDCARD]
